demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c

```diff
@@ -0,0 +1,1608 @@
+        }
+    }
+
+    /* Clean up on error. */
+    if( ( sslStatus == 0 ) && ( pSslSetup != NULL ) )
+    {
+        SSL_free( *pSslContext );
+        *pSslContext = NULL;
+    }
+
+    if( pRootCaFile != NULL )
+    {
+        ( void ) fclose( pRootCaFile );
+    }
+
+    if( pRootCa != NULL )
+    {
+        X509_free( pRootCa );
+    }
+
+    if( pSslSetup != NULL )
+    {
+        SSL_CTX_free( pSslSetup );
+    }
+
+    /* Log failure or success and update the correct exit status to return. */
+    if( sslStatus != 1 )
+    {
+        LogError( ( "Failed to establish a TLS connection to %.*s.",
+                    AWS_ENDPOINT_LENGTH,
+                    AWS_ENDPOINT ) );
+        status = EXIT_FAILURE;
+    }
+    else
+    {
+        LogInfo( ( "Established a TLS connection to %.*s.\n\n",
+                   AWS_ENDPOINT_LENGTH,
+                   AWS_ENDPOINT ) );
+        status = EXIT_SUCCESS;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t transportSend( MQTTNetworkContext_t pSslContext,
+                              const void * pMessage,
+                              size_t bytesToSend )
+{
+    int32_t bytesSent = 0;
+    int pollStatus = 0;
+    struct pollfd fileDescriptor =
+    {
+        .events  = POLLOUT,
+        .revents = 0
+    };
+
+    /* Set the file descriptor for poll. */
+    fileDescriptor.fd = SSL_get_fd( pSslContext );
+
+    /* Poll the file descriptor to check if SSL_Write can be done now. */
+    pollStatus = poll( &fileDescriptor, 1, TRANSPORT_SEND_RECV_TIMEOUT_MS );
+
+    if( pollStatus > 0 )
+    {
+        bytesSent = ( int32_t ) SSL_write( pSslContext, pMessage, bytesToSend );
+    }
+    else if( pollStatus == 0 )
+    {
+        LogDebug( ( "Timed out while polling SSL socket for write buffer availability." ) );
+    }
+    else
+    {
+        LogError( ( "Polling of the SSL socket for write buffer availability failed"
+                    " with status %d.",
+                    pollStatus ) );
+        bytesSent = -1;
+    }
+
+    return bytesSent;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t transportRecv( MQTTNetworkContext_t pSslContext,
+                              void * pBuffer,
+                              size_t bytesToRecv )
+{
+    int32_t bytesReceived = 0;
+    int pollStatus = -1, bytesAvailableToRead = 0;
+    struct pollfd fileDescriptor =
+    {
+        .events  = POLLIN | POLLPRI,
+        .revents = 0
+    };
+
+    /* Set the file descriptor for poll. */
+    fileDescriptor.fd = SSL_get_fd( pSslContext );
+
+    /* Check if there are any pending data available for read. */
+    bytesAvailableToRead = SSL_pending( pSslContext );
+
+    /* Poll only if there is no data available yet to read. */
+    if( bytesAvailableToRead <= 0 )
+    {
+        pollStatus = poll( &fileDescriptor, 1, TRANSPORT_SEND_RECV_TIMEOUT_MS );
+    }
+
+    /* SSL read of data. */
```

Comment:
Update the comment to- "Read data if it was already pending or if it became available during polling. You should change the order of the check here to ensure that pollStatus  is only checked if poll was called. 

```c
 if(( bytesAvailableToRead > 0 ) || ( pollStatus > 0 ))
```